test/functional/p2p_tx_download.py

```diff
@@ -241,6 +244,35 @@ def test_spurious_notfound(self):
         self.log.info('Check that spurious notfound is ignored')
         self.nodes[0].p2ps[0].send_message(msg_notfound(vec=[CInv(MSG_TX, 1)]))
 
+    def test_unrequested_tx(self):
+        self.log.info("Check that nodes don't process unrequested txn")
+        self.restart_node(0)
+
+        unrequested_tx = CTransaction()
+        unrequested_tx.vin.append(CTxIn())
+
+        unrequested_tx.vout.append(CTxOut())
+        unrequested_tx.calc_sha256()
+
+        peer = self.nodes[0].add_p2p_connection(P2PInterface())
+        peer.send_message(msg_tx(unrequested_tx))
+        with self.nodes[0].assert_debug_log(expected_msgs=['unrequested transaction from peer=0']):
+            peer.sync_with_ping()
+            assert_equal(len(self.nodes[0].getrawmempool()), 0)
+        peer.peer_disconnect()
+        peer.wait_for_disconnect()
+
+        self.log.info("Check that nodes process unrequested txn with relay permission ")
+        self.restart_node(0, extra_args=['-whitelist=relay@127.0.0.1'])
+
+        peer = self.nodes[0].add_p2p_connection(P2PInterface())
+        peer.send_message(msg_tx(unrequested_tx))
+        with self.nodes[0].assert_debug_log(expected_msgs=['from peer=0 was not accepted: scriptpubkey']):
+            peer.sync_with_ping()
+            assert_equal(len(self.nodes[0].getrawmempool()), 0)
+        peer.peer_disconnect()
+        peer.wait_for_disconnect()
+
```

Comment:
Seems that `assert_equal(len(self.nodes[0].getrawmempool()), 0)` isn't really doing anything since the tx being used is always invalid so we don't ever expect it to be accepted to the mempool.

Suggest using a valid tx instead and then testing the changes to the mempool directly (count=0 if tx msg is received before being requested and count=1 if tx msg is received after tx is requested) instead of doing an implicit check via the logs using `assert_debug_log`.

Reply:
That's a good suggestion. Initially I was using valid transactions to feed the node and changed it after this [comment](https://github.com/bitcoin/bitcoin/pull/20277#discussion_r517819829). I guess the `getrawmempool` is a leftover. Will remove it on #21224.

I think it's still fine to use invalid transactions as the behavior under check is "Does this transaction reach mempool acceptance ?", which is probed through the `assert_debug_log`. "Is this transaction mempool-valid ?" is already a distinct checked behavior, of which we're not interested here ?